File: chapter14/independent_set/C++/independent_set.cpp
#include <vector>
#include <set>
#include <algorithm>
#include <iostream>
using namespace std;

struct interval {
  int start;
  int end;
  int pos;
};

bool operator < (const interval a, const interval b) {
  return a.end < b.end;
}


void independent_set(int a[], int b[], int n, set<int> &output) {
  vector<interval> items;
  for (int i=0; i<n; i++) {
    interval tmp;
    tmp.start = a[i];
    tmp.end = b[i];
    tmp.pos = i;
    items.push_back(tmp);
  }

  sort(items.begin(), items.end());

  output.insert(items[0].pos);
  int ultimo = 0;

  for (int i=1; i<n; i++) {
    if (items[i].start >= items[ultimo].end) {
      output.insert(items[i].pos);
      ultimo = i;
    }
  }
}


int main() {
  int n = 5;
  int a[] = {0, 3, 2, 9, 0};
  int b[] = {1, 4, 5, 12, 10};

  set<int> output;

  independent_set(a, b, n, output);

  cout << "Result: ";
  for (set<int>::iterator it=output.begin(); it!=output.end(); it++) {
    cout << *it << " ";
  }
  cout << endl;
}

File: chapter14/independent_set/Java/IndependentSet.java
/*
 * Chapter 14
 * Paragraph 2
 *
 * Independent set of intervals, each one has weight 1
 */


import java.util.Collections;
import java.util.ArrayList;
import java.util.Set;
import java.util.TreeSet;


class Interval implements Comparable<Interval> {
  int start;
  int end;
  int pos;

  public int compareTo(Interval b) {
    return end - b.end;
  }
}

public class IndependentSet {
    public static void compute(int a[], int b[], int n, Set<Integer> output) {
        ArrayList<Interval> items = new ArrayList<Interval>();
        for (int i=0; i<n; i++) {
            Interval tmp = new Interval();
            tmp.start = a[i];
            tmp.end = b[i];
            tmp.pos = i;
            items.add(tmp);
        }

        Collections.sort(items);

        output.add(items.get(0).pos);
        int ultimo = 0;

        for (int i=1; i<n; i++) {
            if (items.get(i).start >= items.get(ultimo).end) {
                output.add(items.get(i).pos);
                ultimo = i;
            }
        }
    }

    public static void main(String args[]) {
        int n = 5;
        int a[] = {0, 3, 2, 9, 0};
        int b[] = {1, 4, 5, 12, 10};

        Set<Integer> output = new TreeSet<Integer>();

        IndependentSet.compute(a, b, n, output);

        System.out.print("Result: ");
        for (Integer it : output) {
            System.out.print(it + " ");
        }
        System.out.println("");
    }
}
